AK/StdLibExtraDetails.h

```diff
@@ -475,6 +475,19 @@ using AddRvalueReference = typename __AddReference<T>::RvalueType;
 template<class T>
 requires(IsEnum<T>) using UnderlyingType = __underlying_type(T);
 
+template<typename T, unsigned ExpectedSize, unsigned ActualSize>
+struct __AssertSize : TrueType {
+    static_assert(ActualSize == ExpectedSize,
+        "actual size does not match expected size");
+};
+
+// Note: This type is useful, as the sizes will be visible in the
+//       compiler error messages, as they will be part of the
+//       template parameters. This is not possible with a
+//       static_assert on the sizeof a type.
+template<typename T, unsigned ExpectedSize>
+using AssertSize = __AssertSize<T, ExpectedSize, sizeof(T)>;
```

Comment:
I don't really like how you'd have to `AssertSize<foo, size>::value` to assert this, I can think of two ways, either `inline constexpr bool` for `AssertSize`, or a `consteval` function that will fail (via the same `compiletime_fail` trick used for format string checking) if the sizes aren't correct?

For comparison, here's what `static_assert(AssertSize<int, 5>::value);` produces:
```
/home/Test/Workspace/serenity/./AK/ByteBuffer.h: In instantiation of ‘struct __AssertSize<int, 5, 4>’:
/home/Test/Workspace/serenity/./AK/ByteBuffer.h:28:33:   required from here
/home/Test/Workspace/serenity/./AK/ByteBuffer.h:17:30: error: static assertion failed: actual size does not match expected size
   17 |     static_assert(ActualSize == ExpectedSize,
      |                   ~~~~~~~~~~~^~~~~~~~~~~~~~~
/home/Test/Workspace/serenity/./AK/ByteBuffer.h:17:30: note: ‘(4 == 5)’ evaluates to false
```

what `static_assert(AssertSize<int, 5>);` (with `constexpr bool`) produces:
```
/home/Test/Workspace/serenity/./AK/ByteBuffer.h: In instantiation of ‘struct __AssertSize<int, 5, 4>’:
/home/Test/Workspace/serenity/./AK/ByteBuffer.h:26:78:   required from ‘constexpr const bool AssertSize<int, 5>’
/home/Test/Workspace/serenity/./AK/ByteBuffer.h:28:15:   required from here
/home/Test/Workspace/serenity/./AK/ByteBuffer.h:17:30: error: static assertion failed: actual size does not match expected size
   17 |     static_assert(ActualSize == ExpectedSize,
      |                   ~~~~~~~~~~~^~~~~~~~~~~~~~~
/home/Test/Workspace/serenity/./AK/ByteBuffer.h:17:30: note: ‘(4 == 5)’ evaluates to false
```

and here's what the most informative way I could find to make `static_assert(assert_size<int>(5));` produces:
```
/home/Test/Workspace/serenity/./AK/ByteBuffer.h:46:38:   in ‘constexpr’ expansion of ‘assert_size<int>(__Actual_Size{5}, __Expected_Size{4})’
/home/Test/Workspace/serenity/./AK/ByteBuffer.h:40:25: error: call to non-‘constexpr’ function ‘void compiletime_fail(auto:10) [with auto:10 = const char*]’
   40 |         compiletime_fail("actual size does not match expected size");
      |         ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/Test/Workspace/serenity/./AK/ByteBuffer.h:15:6: note: ‘void compiletime_fail(auto:10) [with auto:10 = const char*]’ declared here
   15 | void compiletime_fail(auto);
      |      ^~~~~~~~~~~~~~~~
```

Reply:
In the interest of transparency I actually did the current and the other two examples you gave before I sent the PR. I decided that I liked the current version the most, as it was the most clear to me and gave the most debug info. I like the conteval / constexpr function, but that doesn't give you the nice compiler error message that the template type versions do.
```
/home/Test/Workspace/serenity/./AK/ByteBuffer.h:17:30: note: ‘(4 == 5)’ evaluates to false
```

So which would you prefer me to do?

Reply:
my only real problem is how the assertion is spelled out, I like the `(4 == 5) evaluates to false` thing too.

Actually, one more option I can think of that probably makes me happy is:
```c++
template<typename T, unsigned ExpectedSize, unsigned ActualSize>
struct __AssertSize : TrueType {
    static_assert(ActualSize == ExpectedSize,
        "actual size does not match expected size");

    consteval operator bool() const { return value; }
};
```

so we can say `static_assert(AssertSize<int, 5>());`.

Reply:
ok, done